atomic_reactor/plugins/pre_reactor_config.py

```diff
@@ -0,0 +1,155 @@
+    """
+    try:
+        workspace = workflow.plugin_workspace[ReactorConfigPlugin.key]
+        return workspace[WORKSPACE_CONF_KEY]
+    except KeyError:
+        # The plugin did not run or was not successful: use defaults
+        conf = ReactorConfig(DEFAULT_CONFIG)
+        workspace = workflow.plugin_workspace.get(ReactorConfigPlugin.key, {})
+        workspace[WORKSPACE_CONF_KEY] = conf
+        workflow.plugin_workspace[ReactorConfigPlugin.key] = workspace
+        return conf
+
+
+class ClusterConfig(object):
+    """
+    Configuration relating to a particular cluster
+    """
+
+    def __init__(self, name, max_concurrent_builds=None, enabled=True):
+        self.name = str(name)
+        self.max_concurrent_builds = None
+        self.enabled = enabled
+
+        if max_concurrent_builds is not None:
+            self.max_concurrent_builds = int(max_concurrent_builds)
+
+
+class ReactorConfig(object):
+    """
+    Class to parse the atomic-reactor configuration file
+    """
+
+    CLUSTERS_KEY = 'clusters'
+
+    def __init__(self, config, log=None):
+        self.conf = config or {}
+        self.log = log or logging.getLogger(__name__)
+
+        # Prepare cluster configurations
+        self.cluster_configs = {}
+        for platform, clusters in self.conf.get(self.CLUSTERS_KEY, {}).items():
+            cluster_configs = [ClusterConfig(**cluster) for cluster in clusters]
+            self.cluster_configs[platform] = [conf for conf in cluster_configs
+                                              if conf.enabled]
+
+    def get_enabled_clusters_for_platform(self, platform):
+        return self.cluster_configs.get(platform, [])
+
+
+class ReactorConfigPlugin(PreBuildPlugin):
+    """
+    Parse atomic-reactor configuration file
+    """
+
+    # Name of this plugin
+    key = 'reactor_config'
+
+    # Exceptions from this plugin should fail the build
+    is_allowed_to_fail = False
+
+    def __init__(self, tasker, workflow, config_path):
+        """
+        constructor
+
+        :param tasker: DockerTasker instance
+        :param workflow: DockerBuildWorkflow instance
+        :param config_path: str, configuration filename
+        """
+        # call parent constructor
+        super(ReactorConfigPlugin, self).__init__(tasker, workflow)
+        self.config_path = config_path
+
+    def run(self):
+        """
+        Run the plugin
+
+        Parse and validate config.
+        Store in workflow workspace for later retrieval.
+        """
+
+        self.log.info("reading config from %s", self.config_path)
+        with open(self.config_path) as fp:
+            conf = yaml.safe_load(fp)
+
+        # Validate against JSON schema
+        try:
+            resource = resource_stream('atomic_reactor', 'schemas/config.json')
+            config_schema = codecs.getreader('utf-8')(resource)
+        except (IOError, TypeError):
+            self.log.error("unable to extract JSON schema, cannot validate")
```

Comment:
In this situation, the schema itself could not be retrieved. Yes, I guess that should be fatal, will fix.